### PR TITLE
[release/7.0] Fix server-side OCSP stapling on Linux

### LINKED ISSUE #96770: SslStreamCertificateContext stops fetching OCSP staples after one request failure
Related to https://github.com/dotnet/runtime/issues/89907. The fix (https://github.com/dotnet/runtime/pull/90200) prevents the invalid OCSP staple from being sent out, but the `_pendingDownload` field is never cleared which prevents the code from fetching another OCSP staple in the future.

### PR DESCRIPTION
**Pull request by rzikm**
Backport of PR #96792, PR #96448, PR #90200 and PR #96972.

Fixes #96770, #96659 and #89907

# Description

Regression: No, .NET 6 didn't have OCSP staple feature.
Customer: Internal partner team - blocking migration from Windows to Linux.

OCSP (Online Certificate Status Protocol) stapling is an optimization where instead of clients individually retrieving revocation status of the server certificate, server will fetch the OCSP response itself and send it to clients during connection handshake. The authenticity of the response is assured by a digital signature.

First bug in .NET 7.0+ ... An invalid response can get cached and the server would fail to refresh it. The server would then keep sending the old, cached and potentially malformed OCSP response. This bug is triggered when either:
- The OCSP response could not be parsed (e.g. due to OCSP server sending an error page due to outage)
- .NET fails to validate the authenticity of the OCSP response
This has been partially fixed in main by PR #90200 and completely fixed in main by PR #96448

Second bug in .NET 7.0+ ... Validation of OCSP response always fails when the method of "delegated signing" is used (delegated signing means that the OCSP response is signed by a special certificate delegated by the server certificate issuer).
- Note that validation when receiving OCSP response on client-side is not affected by this bug and is extensively tested, only the server part is affected by the bug.
- Server validating the OCSP response before sending it out is only to play nice and not send invalid OCSP responses out. This bug DOES NOT create security vulnerability because clients are expected to independently validate the OCSP response themselves.
Fixed in main by PR #96792

The two issues mentioned above may lead to following undesired behaviors:
- Server caches an invalid OCSP staple (e.g. error page due to OCSP server outage), and keeps sending it out.
- Server stops refreshing the OCSP staple and keeps sending the old one even after its validity expired.
- Server sends out OCSP staple which it does not consider valid

# Customer Impact

Android clients cannot connect to .NET 7+ servers affected by this bug because the OCSP information may get outdated (and not refreshed) and Android 9+'s application default security restrictions don't allow HTTP connections by default.

# Regression

No, sending OCSP staples from .NET servers is a new feature in .NET 7.

# Testing

Locally reproduced affected scenario and extensively tested manually.

Customer validated private 7.0 bits.

There is extensive existing test coverage on client-side OCSP usage/validation. Missing E2E server-side automated test coverage is planned for upcoming weeks.

# Risk

Small to medium. Code touched by this PR is not used in other code paths than OCSP, so only "Sending OCSP staples from a server" scenario is affected.

### CODE REVIEW

**Comment by rzikm** [comment]
No *.ref assemblies were touched by this PR

**Comment by carlossanlop** [comment]
There was a generalized failure in the 7.0 branch which impacted this PR. I'm updating the branch so that your PR gets rebased to the latest bits.

**Comment by karelz** [comment]
Approved by Tactics (@SteveMCarroll) on 1/15 via email - label updated to Servicing-approved.

**Comment by carlossanlop** [comment]
@karelz @rzikm who can provide a sign-off? Are the CI failures unrelated?

**Review by wfurt — APPROVED** [review]
LGTM

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStreamCertificateContext.Linux.cs [hunk 1]
```diff
@@ -245,15 +265,28 @@ private SslStreamCertificateContext(X509Certificate2 target, X509Certificate2[]
...
                 GC.KeepAlive(caCert);
+
+                _pendingDownload = null;
+                if (ret == null)
+                {
+                    // all download attempts failed, don't try again for 5 seconds.
+                    // Note that if server does not send OCSP staples, clients may still
+                    // contact OCSP responders directly.
+                    _nextDownload = DateTimeOffset.UtcNow.AddSeconds(5);
+                    _ocspExpiration = _nextDownload;
```
Comment (bartonjs) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStreamCertificateContext.Linux.cs:288]: From the PR/discussion in main, this shouldn't change _ocspExpiration.

Comment (carlossanlop) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslStreamCertificateContext.Linux.cs:288]: Same request as [here](https://github.com/dotnet/runtime/pull/96838#discussion_r1454042186). @rzikm @wfurt can one of you look at this ASAP?

Comment (wfurt) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslStreamCertificateContext.Linux.cs]: updated with #96972

**Review by bartonjs — APPROVED** [review]

**Pull request merged**